Tests/SwiftWarningControlTest/WarningControlTests.swift

```diff
@@ -0,0 +1,272 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftParser
+import SwiftSyntax
+@_spi(ExperimentalLanguageFeatures) import SwiftWarningControl
+import XCTest
+import _SwiftSyntaxGenericTestSupport
+import _SwiftSyntaxTestSupport
+
+public class WarningGroupControlTests: XCTestCase {
+  func testSimpleFunctionWarningGroupControl() throws {
+    try validateWarningGroupControl(
+      """
+      @warn(GroupID, as: error)
+      func foo() {
+          1️⃣let x = 1
+      }
+      """,
+      diagnosticGroupID: "GroupID",
+      states: [
+        "1️⃣": .error
+      ]
+    )
+  }
+
+  func testNestedFunctionWarningGroupControl() throws {
+    try validateWarningGroupControl(
+      """
+      @warn(SomeOtherGroup, as: warning)
+      @warn(GroupID, as: error)
+      @warn(YetAnotherGroup, as: warning)
+      func foo() {
```

Comment:
A little nitpicky but should the code in the test cases be indented by 2 spaces to esthetically match the surrounding Swift style?